api/src/routes/template-v2.ts

```diff
@@ -0,0 +1,185 @@
+import express, { Request, Response, Router } from "express";
+import { body } from "express-validator";
+import { validateRequest } from "../middlewares/validate-request";
+import { TemplateService } from "../services/template";
+import { genHash } from "../utils/gen-hash";
+import { mixpanelTrack } from "../utils/mixpanel";
+import { parseCadence } from "../utils/parse-cadence";
+
+function templateRouter(
+  templateService: TemplateService,
+  namesJSONFile: JSON
+): Router {
+  const router = express.Router();
+
+  router.get("/templates", async (req: Request, res: Response) => {
+    const name = req.query.name as string;
+
+    if (!name) {
+      mixpanelTrack("get_template_by_name", {
+        name,
+        status: 400,
+      });
+
+      res.status(400);
+      return res.send(
+        `GET /templates-- Required query parameter "name" not provided.`
+      );
+    }
+
+    let templateId: string = "";
+    let _name: string = name;
+    while (_name !== undefined) {
+      let foundName = namesJSONFile[_name];
+      if (foundName !== undefined) templateId = foundName;
+      _name = foundName;
+    }
+
+    const template = await templateService.getTemplate(templateId);
+
+    if (!template) {
+      mixpanelTrack("get_template_by_name", {
+        name,
+        templateId,
+        status: 204,
+      });
+
+      res.status(204);
+      return res.send(
+        `GET /templates/:template_id -- Did not find template for template_id=${templateId}`
+      );
+    }
+
+    mixpanelTrack("get_template_by_name", {
+      name,
+      templateId,
+      status: 200,
+    });
+
+    return res.send(template);
+  });
+
+  router.get("/templates/manifest", async (req: Request, res: Response) => {
+    const templateManifest = await templateService.getTemplateManifest();
+
+    if (!templateManifest) {
+      mixpanelTrack("get_template_manifest", {
+        status: 204,
+      });
+
+      res.status(204);
+      return res.send(
+        `GET /templates/manifest -- Did not find template manifest`
+      );
+    }
+
+    mixpanelTrack("get_template_manifest", {
+      status: 200,
+    });
+
+    return res.send(templateManifest);
+  });
+
+  router.get("/templates/:template_id", async (req: Request, res: Response) => {
+    const templateId = req.params.template_id;
+
+    const template = await templateService.getTemplate(templateId);
+
+    if (!template) {
+      mixpanelTrack("get_template", {
+        templateId,
+        status: 204,
+      });
+
+      res.status(204);
+      return res.send(
+        `GET /templates/:template_id -- Did not find template for template_id=${templateId}`
+      );
+    }
+
+    mixpanelTrack("get_template", {
+      templateId,
+      status: 200,
+    });
+
+    return res.send(template);
+  });
+
+  router.post("/templates/search/messages", async (req: Request, res: Response) => {
```

Comment:
Nit: The endpoint path is a bit confusing to me, as it could also be interpreted as searching for a "message" resource. But that's not true, I think it's only searching by "message" fields like title/description.

Could it make more sense for this to just be `/templates/search` or even just `GET /templates?search={query}` (where an empty query or no query parameter present would list all templates)? Using the GET verb could also make sense (semantically at least), since this only reads data.

Reply:
I can see there is also an endpoint for searching by the AST hash. Maybe that could instead just be another query parameter like `/templates/search?cadence_ast_hash`?

It could lead to simpler API if there was a single endpoint that supported different search options, as opposed to multiple endpoints, each supporting a different search option. 

Thoughts?